DigitalLearningSolutions.Web/ControllerHelpers/DateValidator.cs

```diff
@@ -1,172 +1,135 @@
-            error.YearValid = YearIsValid(year);
-            if (
-                !error.DayValid && !error.MonthValid
-                || !error.DayValid && !error.YearValid
-                || !error.MonthValid && !error.YearValid
-                || error.DayValid && error.MonthValid && error.YearValid
-            )
-            {
-                error.DayValid = false;
-                error.MonthValid = false;
-                error.YearValid = false;
-            }
-
-            error.ErrorMessage = ConstructErrorMessage(day, month, year, name);
-
-            return error;
-        }
-
-        private static bool DayIsValid(int day)
-        {
-            return day > 0 && day < 32;
-        }
-
-        private static bool MonthIsValid(int day)
-        {
-            return day > 0 && day < 13;
-        }
-
-        private static bool YearIsValid(int year)
-        {
-            return year > 1752 && year < 10000;
-        }
-
-        private static string ConstructErrorMessage(int day, int month, int year, string name)
-        {
-            List<string> emptyElements = new List<string>();
-
-            if (day == 0)
-            {
-                emptyElements.Add("day");
-            }
-
-            if (month == 0)
-            {
-                emptyElements.Add("month");
-            }
-
-            if (year == 0)
-            {
-                emptyElements.Add("year");
-            }
-
-            if (emptyElements.Any())
-            {
-                return name + " must include a " + string.Join(" and ", emptyElements);
-            }
-
-            return name + " must be a real date";
-        }
-    }
-}
+﻿namespace DigitalLearningSolutions.Web.ControllerHelpers
+{
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel.DataAnnotations;
+
+    public static class DateValidator
+    {
+        public static DateValidationResult ValidateDate(
+            int? day,
+            int? month,
+            int? year,
+            string name = "Date",
+            bool required = false
+        )
+        {
+            if (!day.HasValue && !month.HasValue && !year.HasValue)
+            {
+                return required ? new DateValidationResult(name + " is required") : new DateValidationResult();
+            }
+
+            if (day.HasValue && month.HasValue && year.HasValue)
+            {
+                return ValidateDate(day.Value, month.Value, year.Value, name);
+            }
+
+            var errorMessage = GetMissingValuesErrorMessage(day, month, year, name);
+            return new DateValidationResult(!day.HasValue, !month.HasValue, !year.HasValue, errorMessage);
+        }
+
+        private static DateValidationResult ValidateDate(int day, int month, int year, string name)
+        {
+            var invalidDay = day < 1 || day > 31;
+            var invalidMonth = month < 1 || month > 12;
+            var invalidYear = year < 1;

```

Comment:
Minimum year is handled by the fact that date must be in future. Maximum year gives the 'real date' error.

Reply:
From another comment, year > 10000 now gives a more specific error. Left minimum year as 0 otherwise the error message indicating that 1752 is not a real year is a bit strange.